presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/AbstractResourceConfigurationManager.java

```diff
@@ -104,6 +111,41 @@ protected void validateRootGroups(ManagerSpec managerSpec)
         return selectors.build();
     }
 
+    @Override
+    public Optional<SelectionContext<VariableMap>> match(SelectionCriteria criteria)
+    {
+        return pickSelector(criteria).map(result -> new SelectionContext<>(
+                    result.getResourceGroupIdTemplate().expandTemplate(result.getVariableMap()),
+                    result.getVariableMap(),
+                    result.getResourceGroupIdTemplate().getFirstDynamicSegment(),
+                    getQueryLimits(result.getResourceGroupIdTemplate())));
```

Comment:
This doesn't seem like the right place to put this.  In other words, what do resource group limits have to do with the selection context, which is the resource group ID that is selected, and can we wait to inject this until it hits the resource group?

Reply:
Another area where we thought a lot and I decided on this because 

(1) `SelectionContext` is the part where the `ResourceGroupManager` informs the Coordinator about various details about the decisions taken by the resource group (main being resource group id - but it also contains the `VariableMap` and `firstDynamicSegmentPosition` which are in addition to the `ResourceGroupId`. Now, one of the decisions that the resource group takes is the limits that the resource group expects the coordinator to enforce and they are returned as a part of the `SelectionContext`. And hence, I felt that this is a part of the Selection that the resource group manager makes based on the given `SelectionCriteria`.

(2) This also helps simplify our design a bit. By doing this - we are able to inject the limits object into the appropriate `Query` objects to be enforced by `QueryTracker` and `SqlQueryManager`. The alternate would have been to change the prototype of `ResourceGroupManager#submit` and deal with non-final variables. While this is a very doable change - I didn't see any advantage of going this approach and liked the straight forwardness and simplicity of this approach.

Reply:
The `VariableMap` relates to the `ResourceGroupId`--it's the values that are matched if there's a regular expression in the resource group selector.  `firstDynamicSegmentPosition` is an awkward name and probably a hack, but also relates to the `ResourceGroupId`--it's the first position that's "dynamic" (contains a regex matcher).  So I feel like this method isn't the right place to put this, because we're overloading the intended purpose from "matching a resource group node" to that and pulling out the resource group limits.

Many methods on `ManagedQueryExecution` and `QueryExecution` are stateful, so I think it's fine to statefully add these limits there.  And I think it should be possible to inject these per query limits onto the RG itself, similar to other existing limits that are not per query.  Why not have the RG add the limits once it's submitted to the RG?